### PR TITLE
sunset linalg jazz

### PR DESCRIPTION
**Pull request by Sacha0**
This pull request is the next step towards JuliaLang/LinearAlgebra.jl#57 after JuliaLang/julia#24969, JuliaLang/julia#25083, JuliaLang/julia#25125, and JuliaLang/julia#25148, completing item 18 and part of 19 in JuliaLang/julia#24969's OP's task list. Specifically, this pull request: (1) deprecates all `A[ct]_(mul|ldiv|rdiv)_B[ct][!]` methods in favor of `*`, `/`, `\`, `mul!`, `ldiv!`, and `rdiv!`; (2) completes the transition from `ConjRowVector`/`RowVector` to `Adjoint`/`Transpose` by deprecating the former's constructors with pointers to the latter (while preserving associated functionality for the deprecation period) ; and (3) deprecates `adjoint(v)`/`transpose(v)` for `v<:AbstractVector`, which were already lazy, to `Adjoint(v)`/`Transpose(v)`.

This pull request largely completes the work towards JuliaLang/LinearAlgebra.jl#57. The remaining certain work is a solid news item, documentation, and a bit of cleanup; I plan to wrap those into future pull requests after sorting out other things feature-y. The remaining uncertain work / design question is what to do about `adjoint`/`transpose` for matrices; hopefully triage can sort that point out tomorrow and I will implement the decision in a downstream pull request.

Thanks and best!

### CODE REVIEW

File: base/exports.jl [hunk 1]
```diff
@@ -220,34 +218,6 @@ export
...
-    Ac_rdiv_Bc,
-    At_ldiv_B,
-    At_ldiv_B!,
-    At_ldiv_Bt,
-    At_mul_B,
-    At_mul_B!,
-    At_mul_Bt,
-    At_mul_Bt!,
-    At_rdiv_B,
-    At_rdiv_Bt,
```
Comment (StefanKarpinski) [inline base/exports.jl]: Oh, man. That is satisfying.

File: base/linalg/rowvector.jl [hunk 2]
```diff
@@ -1,5 +1,7 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
+# TODO: remove the type stubs below between 0.7 and 1.0
```
Comment (Sacha0) [inline base/linalg/rowvector.jl]: I left these type stubs in place to preserve these types' concatenation behavior during the deprecation period. (These types' concatenation behavior is defined through a set of large unions including these types in base/sparse/sparsevector.jl, which loads before base/deprecated.jl.)

File: base/sparse/sparsevector.jl [hunk 3]
```diff
@@ -1007,9 +1007,9 @@ const _Triangular_DenseArrays{T,A<:Matrix} = Base.LinAlg.AbstractTriangular{T,A}
...
 
-const _SparseConcatGroup = Union{Vector, Adjoint{<:Any,<:Vector}, Transpose{<:Any,<:Vector}, RowVector{<:Any,<:Vector}, Matrix, _SparseConcatArrays, _Annotated_SparseConcatArrays, _Annotated_DenseArrays}
-const _DenseConcatGroup = Union{Vector, Adjoint{<:Any,<:Vector}, Transpose{<:Any,<:Vector}, RowVector{<:Any, <:Vector}, Matrix, _Annotated_DenseArrays}
-const _TypedDenseConcatGroup{T} = Union{Vector{T}, Adjoint{T,Vector{T}}, Transpose{T,Vector{T}}, RowVector{T,Vector{T}}, Matrix{T}, _Annotated_Typed_DenseArrays{T}}
+const _SparseConcatGroup = Union{Vector, Adjoint{<:Any,<:Vector}, Transpose{<:Any,<:Vector}, Base.LinAlg.RowVector{<:Any,<:Vector}, Matrix, _SparseConcatArrays, _Annotated_SparseConcatArrays, _Annotated_DenseArrays}
+const _DenseConcatGroup = Union{Vector, Adjoint{<:Any,<:Vector}, Transpose{<:Any,<:Vector}, Base.LinAlg.RowVector{<:Any, <:Vector}, Matrix, _Annotated_DenseArrays}
+const _TypedDenseConcatGroup{T} = Union{Vector{T}, Adjoint{T,Vector{T}}, Transpose{T,Vector{T}}, Base.LinAlg.RowVector{T,Vector{T}}, Matrix{T}, _Annotated_Typed_DenseArrays{T}}
```
Comment (Sacha0) [inline base/sparse/sparsevector.jl]: `RowVector` remains in these unions to preserve its concatenation behavior during the deprecation period.

File: test/linalg/rowvector.jl [hunk 4]
```diff
@@ -1,318 +0,0 @@
...
-    v = [1]
-    rv = rvtranspose(v)
-    rv[CartesianIndex()] = 2
-    @test rv[CartesianIndex()] == 2
-    rv[CartesianIndex(1)] = 1
-    @test rv[CartesianIndex(1)] == 1
-
-    # setindex!(v::RowVector, ...) should return v rather than v's parent
-    @test setindex!(RowVector([1, 2, 3]), 4, 1)::RowVector == [4 2 3]
-end
```
Comment (Sacha0) [inline test/linalg/rowvector.jl]: Equivalents of almost all of these tests (and more besides) now exercise `Adjoint`/`Transpose`.

File: base/deprecated.jl [hunk 5]
```diff
@@ -2180,12 +2180,6 @@ end
     @deprecate chol!(x::Number, uplo) chol(x) false
 end
 
-# deprecate RowVector{T}(shape...) constructors to RowVector{T}(uninitialized, shape...) equivalents
-@deprecate RowVector{T}(n::Int) where {T}               RowVector{T}(uninitialized, n)
-@deprecate RowVector{T}(n1::Int, n2::Int) where {T}     RowVector{T}(uninitialized, n1, n2)
-@deprecate RowVector{T}(n::Tuple{Int}) where {T}        RowVector{T}(uninitialized, n)
-@deprecate RowVector{T}(n::Tuple{Int,Int}) where {T}    RowVector{T}(uninitialized, n)
```
Comment (Sacha0) [inline base/deprecated.jl]: Moved below among other things `RowVector`.

**Comment by Sacha0** [comment]
I would like to give tremendous thanks to (among several other parties!) @andreasnoack, @simonbyrne, @mbauman, and @andyferris, each of whom at one point or another built a branch exploring these changes, and also for Andreas's, Simon's, and Andy's greatly helpful direct involvement in preparing this series of pull requests. ❤️!

**Comment by andreasnoack** [comment]
This is really great and an incredible amount of work. Thank you so much for doing it.

**Comment by stevengj** [comment]
If we have `adjoint(v)/transpose(v)` for matrices, shouldn't we leave them for vectors in order to facilitate writing generic code?

**Comment by StefanKarpinski** [comment]
Triage wants to call the function to materialize a final, non-lazy result `collect` consistent with its current meaning to collect all the values in iterators into a concrete vector.

**Comment by Jutho** [comment]
Thanks for all of this. Wasn't the conclusion from #16772 that 2-arg `mul!` methods would get a different name that indicates which argument is overwritten?

**Comment by Sacha0** [comment]
> Wasn't the conclusion from #16772 that 2-arg mul! methods would get a different name that indicates which argument is overwritten?

Separate work :).

**Comment by andyferris** [comment]
To clarify - is `adjoint(::AbstractVector)` only deprecated because it's meaning will change to be create an `Adjoint` instead of a `RowVector`? And we'll reintroduce a method for this in v1.0?

**Comment by andyferris** [comment]
Hmm... I have to say I'm still a bit confused what the plan is here. I guess I might have missed something, sorry! Let me nut this out:

`adjoint` is a perfectly valid mathematical function (like `conj`) and IMO clearly should stay. We need that `adjoint(adjoint(v))` is a no-op (functionally speaking - it might e.g. return a copy of `v`) and that `Adjoint(Adjoint(v))` constructs an `Adjoint`. Also, the outermost `Adjoint(...)` should definitely be the (very shallow?) copy-constructor `Adjoint(::Adjoint)`, if `Adjoint` is a type. So `Adjoint(Adjoint(v)) == Adjoint(v)` because that's just how types and constructors work in Julia, right? Otherwise I'd find that super confusing. So, as I see it, we can't have `Adjoint` being the lazy but otherwise functionally equivalent replacement of `adjoint`.

I definitely agree that it's messy when some methods of a function are eager and some are lazy - but I thought the plan here had been to make all `transpose` and `adjoint` calls lazy (and document that)? I would advocate for that, anyway (it's really easy to use `copy`, after all). I feel that the most breaking thing here (in this series of PRs) will be that `transpose` and `adjoint` of matrices will no longer be copying - the renaming of `RowVector` seems pretty cosmetic.

**Comment by Sacha0** [comment]
Much thanks for the thoughtful and cogent post Andy! I lack strong sentiments regarding which name should have `Adjoint`'s present semantics. (Minimizing breakage was the primary reason for `Adjoint` having its present semantics, but at this point what's another grain of sand on the beach 😉.) If general sentiment favors giving `Adjoint`'s present semantics to `adjoint` and accepting the additional breakage, I would be delighted to drop this pull request's last commit / downstream work in progress and pivot to the plan just above. Thoughts all? In the interest of seeing this and downstream work through in a timely manner, rapid formation of a quorum either way would be lovely.

**Comment by jekbradbury** [comment]
I'm developing a new array-like wrapper type over at https://github.com/jekbradbury/Minibatch.jl, and I'm coming at this as a relative beginner. From that perspective I think there are some things that might be worth mentioning:

- It would be nice if there were a single, obvious level of abstraction where I should add multiplication methods (i.e., either to `Base.:*` or to `LinAlg.mul(!)`). Looks like there is with this proposal (the latter) but not currently (I could add methods for all the jazzy functions but it's easier to let them use fallbacks and just add methods to `*`).
- I'd like to be able to indicate that I want the lazy adjoint wrapper to "pass through" my type. So if `a isa MyType{Matrix{Float64}}` then `a'` should have type `MyType{Adjoint{Matrix{Float64}}}` rather than `Adjoint{MyType{Matrix{Float64}}}`, since the `Adjoint` wrapper shouldn't have to know how to deal with my type, while my type knows exactly how to deal with an adjoint wrapper (pass it on to the `mul` functions we're calling). This suggests that lowering `a'` directly to `Adjoint(a)` is not ideal, while lowering it to `adjoint(a)` (with a fallback method `adjoint(a::AbstractArray) = Adjoint(a)`) would work very cleanly.

**Comment by Sacha0** [comment]
Rebased and dropped the last commit, leaving only uncontroversial changes in this pull request. Absent objections or requests for time, I plan to merge these changes tomorrow evening PT at the earliest.

With the support above for transferring `Adjoint`/`Transpose`'s semantics to `adjoint`/`transpose`, and additional support voiced on slack, I nixed existing work in progress and instead am pursuing the preceding semantics transfer alongside introduction of `collect` for materialization. Thanks all! :)

**Comment by Sacha0** [comment]
(Marked triage for merge check.)

**Comment by ararslan** [comment]
IT'S HAPPENING!!!! :tada:

**Pull request merged**